### PR TITLE
[POAE7-2653] support time type/date add func in nextgen

### PR DESCRIPTION
**Pull request by kd97**
### What changes were proposed in this pull request?

- add date/time type
- cast between date and time support
- redefine date/time runtime functions, move more branch to the compile phase.

### Why are the changes needed?

Nextgen feature migration

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs
### Which label does this PR belong to?

CODEREFACTOR

### CODE REVIEW

File: cider/type/plan/DateExpr.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+std::shared_ptr<Analyzer::Expr> DateaddExpr::rewrite_with_child_targetlist(
+    const std::vector<std::shared_ptr<TargetEntry>>& tlist) const {
+  return makeExpr<DateaddExpr>(type_info,
+                               field_,
+                               number_->rewrite_with_child_targetlist(tlist),
+                               datetime_->rewrite_with_child_targetlist(tlist));
+}
+
+std::shared_ptr<Analyzer::Expr> DateaddExpr::rewrite_agg_to_var(
```
Comment (yma11) [inline cider/type/plan/DateExpr.cpp:117]: For those legacy methods of `DateAddExpr`, can you add comments for the ones which we should remove in future?

File: cider/type/plan/UnaryExpr.cpp [hunk 2]
```diff
@@ -24,6 +24,52 @@
...
+  int64_t dim_scaled = DateTimeUtils::get_timestamp_precision_scale(
+      abs(operand_ti.get_dimension() - ti.get_dimension()));
+  JITValuePointer cast_scaled =
+      func.createConstant(JITTypeTag::INT64, dim_scaled * kSecondsInOneDay);
+  if (target_width == operand_width) {
+    return JITValuePointer(&operand_val);
+  } else if (target_width > operand_width) {
+    JITValuePointer cast_val =
+        operand_val.castJITValuePrimitiveType(getJITTypeTag(ti.get_type()));
+    return JITValuePointer(cast_val * cast_scaled);
```
Comment (yma11) [inline cider/type/plan/UnaryExpr.cpp:42]: Do we need handle `NULL` property for cast?

Comment (kd97) [reply cider/type/plan/UnaryExpr.cpp:42]: handle null value processing outside in `codegenCastFunc` func,  and have passed null value testcases in `CiderDateNextgenTest`

**Comment by yma11** [comment]
@kd97 Please rebase this PR.

**Comment by kd97** [comment]
@bigPYJ1151 @yma11 pls help review again.

File: src/cider/tests/nextgen/functional/CiderDateNextgenTest.cpp [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <gtest/gtest.h>
+#include "exec/nextgen/Nextgen.h"
+#include "exec/plan/parser/TypeUtils.h"
+#include "tests/utils/CiderTestBase.h"
+
+using namespace cider::exec::nextgen;
```
Comment (yma11) [inline src/cider/tests/nextgen/functional/CiderDateNextgenTest.cpp:27]: As cheng's suggestion, let's simplify the test file name to `DateTest.cpp` when you do the rebase.

**Review by bigPYJ1151 — APPROVED** [review]
Overall looks good.

**Review by yma11 — APPROVED** [review]

**Review by bigPYJ1151 — APPROVED** [review]

**Pull request merged**